### PR TITLE
CRM-20410 Reformat CRM_SMS_BAO_Provider to use standard create function style rather than the split

### PR DESCRIPTION
**Pull request by seamuslee001**
* [CRM-20410: Refactor SMS provider BAO to have proper create function](https://issues.civicrm.org/jira/browse/CRM-20410)

### CODE REVIEW

File: CRM/SMS/BAO/Provider.php [hunk 1]
```diff
@@ -88,30 +88,41 @@ public static function getProviders($selectArr = NULL, $filter = NULL, $getActiv
...
+   * @return array saved values
    */
-  public static function saveRecord($values) {
-    $values['domain_id'] = CRM_Utils_Array::value('domain_id', $values, CRM_Core_Config::domainID());
-    $dao = new CRM_SMS_DAO_Provider();
-    $dao->copyValues($values);
-    $dao->save();
-  }
+  public static function create($params, $ids = array()) {
+    $id = CRM_Utils_Array::value('id', $ids, CRM_Utils_Array::value('id', $params));
```
Comment (eileenmcnaughton) [inline CRM/SMS/BAO/Provider.php]: the $ids array is legacy - just use params for an new fn

File: tests/phpunit/CRM/SMS/BAO/ProviderTest.php [hunk 2]
```diff
@@ -64,12 +64,12 @@ public function testCreateAndUpdateProvider() {
       'is_active' => 1,
       'api_type' => 1,
     );
-    CRM_SMS_BAO_Provider::saveRecord($values);
+    CRM_SMS_BAO_Provider::create($values, array());
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/SMS/BAO/ProviderTest.php]: I tend to think it only matters to test the api - the BAO is tested in the process & the api is the recommended way to call it. BAO can change, but api is the one we commit to

File: CRM/SMS/BAO/Provider.php [hunk 3]
```diff
@@ -88,30 +88,41 @@ public static function getProviders($selectArr = NULL, $filter = NULL, $getActiv
...
+    else {
+      CRM_Utils_Hook::pre('create', 'SmsProvider', NULL, $params);
+    }
+
+    $provider = new static();
+    if ($id) {
+      $provider->id = $id;
+      $provider->find(TRUE);
+    }
+    if ($id) {
```
Comment (eileenmcnaughton) [inline CRM/SMS/BAO/Provider.php]: so, in theory all of this could be in the extension & be removed from here

File: CRM/SMS/Form/Provider.php [hunk 4]
```diff
@@ -171,11 +171,14 @@ public function postProcess() {
     $recData['is_active'] = CRM_Utils_Array::value('is_active', $recData, 0);
     $recData['is_default'] = CRM_Utils_Array::value('is_default', $recData, 0);
 
-    if ($this->_action & CRM_Core_Action::UPDATE) {
-      CRM_SMS_BAO_Provider::updateRecord($recData, $this->_id);
-    }
-    elseif ($this->_action & CRM_Core_Action::ADD) {
-      CRM_SMS_BAO_Provider::saveRecord($recData);
+    if ($this->_action && (CRM_Core_Action::UPDATE || CRM_Core_Action::ADD)) {
```
Comment (eileenmcnaughton) [inline CRM/SMS/Form/Provider.php]: $ids is not required - set $params['id'] instead

File: CRM/SMS/Form/Provider.php [hunk 5]
```diff
@@ -171,11 +171,14 @@ public function postProcess() {
...
+      if (!empty($this->_id)) {
+        $ids = array('id' => $this->_id);
+      }
+      else {
+        $ids = array();
+      }
+      CRM_SMS_BAO_Provider::create($recData, $ids);
```
Comment (eileenmcnaughton) [inline CRM/SMS/Form/Provider.php]: call api not bao

File: CRM/SMS/Form/Provider.php (hunk 4 above)
Comment (seamuslee001) [inline CRM/SMS/Form/Provider.php]: is it not? it seems like it is needed as per basic_create function

Comment (eileenmcnaughton) [inline CRM/SMS/Form/Provider.php]: no - only $params is needed - $ids is being phased out

File: CRM/SMS/BAO/Provider.php (hunk 1 above)
Comment (seamuslee001) [inline CRM/SMS/BAO/Provider.php]: seems like as per https://github.com/civicrm/civicrm-core/blob/master/api/v3/utils.php#L1281 it expects a 2nd param which is ids

File: CRM/SMS/Form/Provider.php (hunk 4 above)
Comment (eileenmcnaughton) [inline CRM/SMS/Form/Provider.php]: (maybe you should comment whereever you got that from...)

Comment (seamuslee001) [inline CRM/SMS/Form/Provider.php]: see https://github.com/civicrm/civicrm-core/blob/master/api/v3/utils.php#L1281

File: CRM/SMS/BAO/Provider.php (hunk 3 above)
Comment (seamuslee001) [inline CRM/SMS/BAO/Provider.php]: up to L116 but yes agree with that if needed

File: CRM/SMS/BAO/Provider.php (hunk 1 above)
Comment (eileenmcnaughton) [inline CRM/SMS/BAO/Provider.php]: I think it handles rather than expects $ids

**Comment by eileenmcnaughton** [comment]
@seamuslee001 this looks good - only question is whether we should put the domain part in an extension. I'm trying to decide if putting in the extension instead of here will make it easier later or it's a 'might as well merge & when @nganivet does his big change it can be adjusted then....

**Comment by nganivet** [comment]
I would say merge ... I always dream I can do so many things, but the hash reality of customer commitments kicks in and I have to postpone. Might be a while before I have time for this big refactor, although it's nice to see these pieces fall in place.

**Comment by eileenmcnaughton** [comment]
Well at least @seamuslee001 effort adding an api will make it easier

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by eileenmcnaughton** [comment]
Oh - I thought we'd merged this when I blamed it for breakage :-)

**Comment by eileenmcnaughton** [comment]
test this please

**Pull request merged**